pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -511,7 +511,9 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
 				set.Name,
 				replicas[target].Name)
 			err := ssc.podControl.DeleteStatefulPod(set, replicas[target])
-			status.CurrentReplicas--
+			if getPodRevision(replicas[target]) == currentRevision.Name {
+				status.CurrentReplicas--
```

Comment:
This seems like an anti-pattern. Kubernetes should be level-triggered: https://speakerdeck.com/thockin/edge-vs-level-triggered-logic

Status should be updated by observing current state, e.g. listing Pods.